devdoc/Metrics.rst

```diff
@@ -0,0 +1,282 @@
+  cluster demands and load. Metrics provide the monitoring capability to react and take
+  action
+
+* Fault diagnosis and resource monitoring
+
+  Metrics provide historical data useful in diagnosing problems by profiling how demand and
+  usage patterns may change prior to a fault. Predictive analysis can also be applied.
+
+* Analysis of jobs/workunits and profiling
+
+  With proper instrumentation, a robust dynamic metric strategy can track workunit
+  processing.
+
+The document consists of several sections in order to provide requirements as well as
+the design of framework components.
+
+Definitions
+===============
+Some definitions are useful.
+
+Metric
+  A measurement defined by a component that represent an internal state that is useful in a system reliability engineering function. In the context
+  of the framework, a metric is an object representing the above.
+
+Metric Value
+  The current value of a metric. A corollary is a metric measurement which represents a view of a metric value that may be based on state.
+
+Collection
+  The process of choosing the relevant metrics and retrieving their values.
+
+Trigger
+  What causes the collection of metric values.
+
+Metric Report
+  Formatted metric values ready for ingestion by a collection system.
+
+
+
+*************
+Use Scenarios
+*************
+This section describes how components expect to use the framework. It is not a complete list of all requirements
+but rather a sample.
+
+
+Roxie
+=====
+Roxie desires to keep a count of many different internal values. Some examples are
+
+* Disk type operations such as seeks and reads
+* Execution totals
+
+  Need to track items such as total numbers of items such as success and failures
+  as well as breaking some counts into individual reasons. For example, failures
+  may need be categorized such as as
+
+  * Busy
+  * Timeout
+  * Bad input
+
+  Or even by priority (high, low, sla)
+
+* Current operational levels such as the length of internal queues
+* The latency of operations such as queue results, agent responses, and gateway responses
+
+Roxie also has the need to track internal memory usage beyond the pod/system level capabilities. Tracking the state of its large fixed memory pool is necessary.
+
+The Roxie buddy system also must track how often and who is completing requests. The "I Beat You To It" set of metrics must be collected and exposed in order to detect pending node failure. While specific action on these counts is not known up front, it appears clear that these values are useful and should be collected.
+
+There does not appear to be a need for creating and destroying metrics dynamically. The set of metrics is most likely to be created at startup and remain active through the life of the Roxie. If, however, stats collection seeps into the metrics framework, dynamic creation and destruction of stats metrics is a likely requirement.
```

Comment:
This is debatable.  I suspect operations will want to monitor deployed queries individually, and if so that means metrics will need to be created and destroyed dynamically - as queries are deployed.  It will not change very frequently on a production system.

Reply:
My statement on the need for creating and destroying metrics was from the Roxie point of view based on comments from Richard. I suppose it's based on the internal operation of Roxie, not the logical use of Roxie. I do believe the need for dynamic metrics exists and should be included in the design.